### PR TITLE
WIP: Harvester load balancer enhancement

### PR DESCRIPTION
**Pull request by yaocw2020**
- Add ipam module(crd, controller, webhook)
- Define two types of LB, one for the VM workload, the other for the guest cluster
- Support changing the IPAM mode
- Add load balancer converter webhook

Related issue:
https://github.com/harvester/harvester/issues/1762
https://github.com/harvester/harvester/issues/1580

design document:
https://github.com/harvester/harvester/pull/1913

Dependency:
https://github.com/harvester/harvester/pull/3776
https://github.com/harvester/webhook/pull/3

### CODE REVIEW

File: pkg/controller/vmi/controller.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		if lb.Spec.WorkloadType == lbv1.Cluster {
+			continue
+		}
+
+		selector, err := utils.NewSelector(lb.Spec.BackendServerSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		// add or update the backend server to the matched load balancer
```
Comment (w13915984028) [inline pkg/controller/vmi/controller.go:66]: vmi changes frequently; for most of the time, it is not related to lb
the lb checks vmi via iterating endpointslice,  each time

it seems need a fast way,  e.g. lb looks record info of vmi (e.g. id, ip), to have a one step match check

File: pkg/controller/ippool/controller.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	return nil
+}
+
+func (h *Handler) OnChange(key string, ipPool *lbv1.IPPool) (*lbv1.IPPool, error) {
+	if ipPool == nil || ipPool.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if ipPool.Spec.Ranges == nil || len(ipPool.Spec.Ranges) == 0 {
+		return nil, fmt.Errorf("invalid ip pool, range could not be empty")
```
Comment (w13915984028) [inline pkg/controller/ippool/controller.go:50]: As a controller, this file lacks of essential log

**Comment by yaocw2020** [comment]
Need to update vendor after the dependency PRs merged.

**Comment by yaocw2020** [comment]
close because replaced by PR https://github.com/harvester/load-balancer-harvester/pull/13.

**Pull request closed without merging**